### ISSUE TITLE
remotewrite vmagent

### ISSUE THREAD

**Issue by zadros4psa** [VictoriaMetrics/VictoriaMetrics#4149]
### Is your question request related to a specific component?

vmagent

### Describe the question in detail

I'm trying to setup the remoteWrite.useVmProto=true flag when deploying the latest vmagent docker image but i keep getting an errror.

```
vmagent:
    container_name: vmagent
    image: victoriametrics/vmagent:latest
    command:
      - "--promscrape.config=/etc/prometheus/prometheus.yml"
      - "--remoteWrite.url=http://victoriametrics:8428/api/v1/write"
      - "--remoteWrite.useVMProto=true"


flag provided but not defined: -remoteWrite.useVMProto.
```

It there anything that i'm missing?

### Troubleshooting docs

- [ ] General - https://docs.victoriametrics.com/Troubleshooting.html
- [ ] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by denisgolius** [comment]
> ### Is your question request related to a specific component?
> 
> vmagent
> ### Describe the question in detail
> 
> I'm trying to setup the remoteWrite.useVmProto=true flag when deploying the latest vmagent docker image but i keep getting an errror.
> 
> ```
> vmagent:
>     container_name: vmagent
>     image: victoriametrics/vmagent:latest
>     command:
>       - "--promscrape.config=/etc/prometheus/prometheus.yml"
>       - "--remoteWrite.url=http://victoriametrics:8428/api/v1/write"
>       - "--remoteWrite.useVMProto=true"
> 
> 
> flag provided but not defined: -remoteWrite.useVMProto.
> ```
> 
> It there anything that i'm missing?
> ### Troubleshooting docs
> 
>     * [ ]  General - https://docs.victoriametrics.com/Troubleshooting.html
> 
>     * [ ]  vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
> 
>     * [ ]  vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

Hi @zadros4psa !
You don't need to set this flag because VictoriaMetrics try to use VMProto by default and if the remote receiver don't support this protocol then vmagent will switch to prometheus remote write protocol.

**Comment by zadros4psa** [comment]
But i'm getting the same message on this config as well
```
- "--remoteWrite.vmProtoCompressLevel=2"
flag provided but not defined: -remoteWrite.vmProtoCompressLevel

**Comment by denisgolius** [comment]
> But i'm getting the same message on this config as well
> 
> ```
> - "--remoteWrite.vmProtoCompressLevel=2"
> flag provided but not defined: -remoteWrite.vmProtoCompressLevel
> ```

We do not recommend to use latest tag. Could your try container image with this tag `victoriametrics/vmagent:v1.90.0` ?

**Comment by denisgolius** [comment]
> Indeed with that image it works, I did not find any specific documentation about the levels of compression.
> 
> Are there a certain number of levels for compression?
> 
> Thanks.

@valyala

**Comment by gowthamsadasivam** [comment]
@zadros4psa -- I was trying to explore the compression levels as well. 
**TLDR;** the compression level value ranges from `0 to 22`.

I also couldn't find much about it in the documentation. Current documentation specifies:

```
-remoteWrite.vmProtoCompressLevel int
     The compression level for VictoriaMetrics remote write protocol. Higher values reduce network traffic at the cost of higher CPU usage. Negative values reduce CPU usage at the cost of increased network traffic. See https://docs.victoriametrics.com/vmagent.html#victoriametrics-remote-write-protocol
```

I tried to explore the source code to find out the supported compression value range.

1. In the file: [pendingseries.go](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmagent/remotewrite/pendingseries.go#L27) I found that the default value is zero.
2. The module `github.com/VictoriaMetrics/VictoriaMetrics/lib/encoding/zstd` uses different `zstd` module depending on whether cgo is enabled or not during the binary compilation.
    1. If **CGO is enabled** - `github.com/valyala/gozstd` module will be used.
    2. If **CGO is NOT enabled** (Ex: Alpine container releases) - `github.com/klauspost/compress/zstd` module will be used.
3. The module `github.com/valyala/gozstd` is a go wrapper for `zstd`. 
    - From the file `https://github.com/valyala/gozstd/blob/master/zstd.h#L81`  has the comment:
      ```
      The library supports regular compression levels from 1 up to ZSTD_maxCLevel(),
      which is currently 22. Levels >= 20, labeled `--ultra`, should be used with
      caution, as they require more memory. The library also offers negative
      compression levels, which extend the range of speed vs. ratio preferences.
      The lower the level, the faster the speed (at the cost of compression).
      ```
   -  The unit test file has the testing range of 0 to 20 as per the file 
      `https://github.com/valyala/gozstd/blob/master/stream_test.go#L66`
   - From the module constant values: `const DefaultCompressionLevel = 3 // Obtained from ZSTD_CLEVEL_DEFAULT.` So we can consider default compression level is `3` .
4. The module `github.com/klauspost/compress/zstd` is a pure go implementation of `zstd` algorithm.
     - According to the official readme file `https://github.com/klauspost/compress/tree/master/zstd` : 
       ```
       The "Default" compression ratio is roughly equivalent to zstd level 3 (default).
       ```
     - From the file `https://github.com/klauspost/compress/blob/master/zstd/encoder_options.go#L192` we can see that the compression level is automatically calculated and set dynamically depending on the user defined level.
        ```
        switch {
       case level < 3:
           return SpeedFastest
       case level >= 3 && level < 6:
           return SpeedDefault
       case level >= 6 && level < 10:
           return SpeedBetterCompression
       default:
           return SpeedBestCompression
        }
        ```
     - Available compression levels for this module. The actual compression level value set by the user is never used.
       ```
       // SpeedFastest will choose the fastest reasonable compression.
       // This is roughly equivalent to the fastest Zstandard mode.
       SpeedFastest
       
        // SpeedDefault is the default "pretty fast" compression option.
        // This is roughly equivalent to the default Zstandard mode (level 3).
        SpeedDefault
        
         // SpeedBetterCompression will yield better compression than the default.
        // Currently it is about zstd level 7-8 with ~ 2x-3x the default CPU usage.
        // By using this, notice that CPU usage may go up in the future.
        SpeedBetterCompression
        
        // SpeedBestCompression will choose the best available compression option.
        // This will offer the best compression no matter the CPU cost.
        SpeedBestCompression
        
         // speedLast should be kept as the last actual compression option.
         // The is not for external usage, but is used to keep track of the valid options.
         speedLast
       ```
5. **To conclude,**
     - If your binary is **CGO** enabled, the default value is `0` (by vmagent wrapper module) and the compression levels can be ranged from `0 to 22` - the standard `zstd` compression level range.
     - If your binary is **NOT CGO** enabled, the default value is `0` (by vmagent wrapper module) and the compression level can be ranged from `0 to 22`, **but the actual value will be determined dynamically by the library and user input will not be considered directly.**

@valyala / @denisgolius please correct me if the above information is wrong.

**Comment by jiekun** [comment]
Close as answered. Many thanks to @gowthamsadasivam
